### PR TITLE
chore(gatsby): convert build to typescript

### PR DESCRIPTION
**Pull request by devrchancay**
## Description

Port src/commands/build.js to TypeScript

## Related Issues
Related to #21995

### CODE REVIEW

File: packages/gatsby/src/commands/build.ts [hunk 1]
```diff
@@ -35,16 +35,16 @@ if (process.env.GATSBY_EXPERIMENTAL_PAGE_BUILD_ON_DATA_CHANGES) {
...
-type BuildArgs = {
-  directory: string,
-  sitePackageJson: object,
-  prefixPaths: boolean,
-  noUglify: boolean,
-  profile: boolean,
-  openTracingConfigFile: string,
+interface IBuildArgs extends IProgram {
+  directory: string
+  sitePackageJson: object
```
Comment (ascorbic) [inline packages/gatsby/src/commands/build.ts:40]: There are typings for `PackageJson` in index.d.ts, which would be better to use here.

File: packages/gatsby/src/commands/build.ts [hunk 2]
```diff
@@ -95,12 +95,14 @@ module.exports = async function build(program: BuildArgs) {
...
-    err => {
-      activity.panic(structureWebpackErrors(`build-javascript`, err))
-    }
-  )
-  activity.end()
+  let stats
+  try {
+    stats = await buildProductionBundle(program, activity.span)
+  } catch (err) {
+    activity.panic(structureWebpackErrors(`build-javascript`, err))
```
Comment (ascorbic) [inline packages/gatsby/src/commands/build.ts:102]: Could you change the string to `Stage.BuildJavascript`

File: packages/gatsby/src/commands/build.ts [hunk 3]
```diff
@@ -191,7 +193,7 @@ module.exports = async function build(program: BuildArgs) {
   try {
     await buildHTML({
       program,
-      stage: `build-html`,
+      stage: BuildHTMLStage.BuildHTML,
```
Comment (ascorbic) [inline packages/gatsby/src/commands/build.ts:196]: `BuildHTMLStage` is deprecated. Can you use `Stage` instead please

File: packages/gatsby/src/commands/build.ts [hunk 4]
```diff
@@ -23,8 +22,9 @@ import {
   showFeedbackRequest,
 } from "../utils/feedback"
 import * as buildUtils from "./build-utils"
-const { boundActionCreators } = require(`../redux/actions`)
+import { boundActionCreators } from "../redux/actions"
```
Comment (ascorbic) [inline packages/gatsby/src/commands/build.ts:25]: As we're importing this, we can probably remove the `require` on line 159: https://github.com/gatsbyjs/gatsby/pull/23695/files#diff-636b7640c2360d47c928027c70960a65L157 (though please test this!)

**Comment by devrchancay** [comment]
@ascorbit thanks for your comments, could you check again?

**Pull request merged**